{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "51736a96",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from glob import glob\n",
    "import processor\n",
    "import bioframe as bf\n",
    "import numpy as np\n",
    "import bioframe.vis\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7b4738c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = '/scratch/groups/horence/angelika/EM_expression/data/'\n",
    "processed_path = './processed_files/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c73149c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/scratch/groups/horence/angelika/EM_expression/data/iso',\n",
       " '/scratch/groups/horence/angelika/EM_expression/data/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_exon_reads.parquet',\n",
       " '/scratch/groups/horence/angelika/EM_expression/data/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_reads.gct.gz',\n",
       " '/scratch/groups/horence/angelika/EM_expression/data/gencode.v38.annotation.gff3.gz',\n",
       " '/scratch/groups/horence/angelika/EM_expression/data/GTEx_Analysis_2017-06-05_v8_STARv2.5.3a_junctions.gct.gz']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glob(data_path+'*')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d7d41ae3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000223972.5;gene_id=ENSG00000223972.5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000456328.2;Parent=ENSG00000223972.5;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:1;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:2;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:3;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom  source     feature  start    end score strand frame  \\\n",
       "0  chr1  HAVANA        gene  11869  14409     .      +     .   \n",
       "1  chr1  HAVANA  transcript  11869  14409     .      +     .   \n",
       "2  chr1  HAVANA        exon  11869  12227     .      +     .   \n",
       "3  chr1  HAVANA        exon  12613  12721     .      +     .   \n",
       "4  chr1  HAVANA        exon  13221  14409     .      +     .   \n",
       "\n",
       "                                          attributes  \n",
       "0  ID=ENSG00000223972.5;gene_id=ENSG00000223972.5...  \n",
       "1  ID=ENST00000456328.2;Parent=ENSG00000223972.5;...  \n",
       "2  ID=exon:ENST00000456328.2:1;Parent=ENST0000045...  \n",
       "3  ID=exon:ENST00000456328.2:2;Parent=ENST0000045...  \n",
       "4  ID=exon:ENST00000456328.2:3;Parent=ENST0000045...  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genecode_path = data_path+'gencode.v38.annotation.gff3.gz'\n",
    "genecode = bf.read_table(genecode_path, schema='gff', skiprows=7)\n",
    "genecode = genecode[~genecode.start.isna()]\n",
    "genecode.start = genecode.start.astype(int)\n",
    "genecode.end = genecode.end.astype(int)\n",
    "\n",
    "genecode.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ce209704",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['./processed_files/genes_filtered.tsv']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glob(processed_path+'*')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "60a75918",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = pd.read_csv(processed_path+'genes_filtered.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "718dbab2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "      <th>Description</th>\n",
       "      <th>Name</th>\n",
       "      <th>Parent</th>\n",
       "      <th>counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>141474</td>\n",
       "      <td>173862</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000241860.7;gene_id=ENSG00000241860.7...</td>\n",
       "      <td>RP11-34P13.13</td>\n",
       "      <td>ENSG00000241860.7</td>\n",
       "      <td>ENSG00000241860.7</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>257864</td>\n",
       "      <td>359681</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000228463.10;gene_id=ENSG00000228463....</td>\n",
       "      <td>AP006222.1</td>\n",
       "      <td>ENSG00000228463.10</td>\n",
       "      <td>ENSG00000228463.10</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>365389</td>\n",
       "      <td>522928</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000237094.12;gene_id=ENSG00000237094....</td>\n",
       "      <td>RP4-669L17.4</td>\n",
       "      <td>ENSG00000237094.12</td>\n",
       "      <td>ENSG00000237094.12</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>586071</td>\n",
       "      <td>827796</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000230021.10;gene_id=ENSG00000230021....</td>\n",
       "      <td>RP11-206L10.17</td>\n",
       "      <td>ENSG00000230021.10</td>\n",
       "      <td>ENSG00000230021.10</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>778747</td>\n",
       "      <td>810065</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000237491.10;gene_id=ENSG00000237491....</td>\n",
       "      <td>LINC01409</td>\n",
       "      <td>ENSG00000237491.10</td>\n",
       "      <td>ENSG00000237491.10</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom  source feature   start     end score strand frame  \\\n",
       "0  chr1  HAVANA    gene  141474  173862     .      -     .   \n",
       "1  chr1  HAVANA    gene  257864  359681     .      -     .   \n",
       "2  chr1  HAVANA    gene  365389  522928     .      -     .   \n",
       "3  chr1  HAVANA    gene  586071  827796     .      -     .   \n",
       "4  chr1  HAVANA    gene  778747  810065     .      +     .   \n",
       "\n",
       "                                          attributes     Description  \\\n",
       "0  ID=ENSG00000241860.7;gene_id=ENSG00000241860.7...   RP11-34P13.13   \n",
       "1  ID=ENSG00000228463.10;gene_id=ENSG00000228463....      AP006222.1   \n",
       "2  ID=ENSG00000237094.12;gene_id=ENSG00000237094....    RP4-669L17.4   \n",
       "3  ID=ENSG00000230021.10;gene_id=ENSG00000230021....  RP11-206L10.17   \n",
       "4  ID=ENSG00000237491.10;gene_id=ENSG00000237491....       LINC01409   \n",
       "\n",
       "                 Name              Parent  counts  \n",
       "0   ENSG00000241860.7   ENSG00000241860.7       6  \n",
       "1  ENSG00000228463.10  ENSG00000228463.10       6  \n",
       "2  ENSG00000237094.12  ENSG00000237094.12      25  \n",
       "3  ENSG00000230021.10  ENSG00000230021.10      15  \n",
       "4  ENSG00000237491.10  ENSG00000237491.10      32  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "876d6a28",
   "metadata": {},
   "outputs": [],
   "source": [
    "transcripts = processor.get_feature(genecode, 'transcript')\n",
    "transcripts = pd.merge(transcripts, genes.Name, left_on='Parent', right_on='Name').drop('Name', axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "acad3167",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "      <th>Parent</th>\n",
       "      <th>tID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>141474</td>\n",
       "      <td>149707</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000484859.1;Parent=ENSG00000241860.7;...</td>\n",
       "      <td>ENSG00000241860.7</td>\n",
       "      <td>ENST00000484859.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>142808</td>\n",
       "      <td>146831</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000490997.5;Parent=ENSG00000241860.7;...</td>\n",
       "      <td>ENSG00000241860.7</td>\n",
       "      <td>ENST00000490997.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>146386</td>\n",
       "      <td>173862</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000466557.6;Parent=ENSG00000241860.7;...</td>\n",
       "      <td>ENSG00000241860.7</td>\n",
       "      <td>ENST00000466557.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>165491</td>\n",
       "      <td>169210</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000662089.1;Parent=ENSG00000241860.7;...</td>\n",
       "      <td>ENSG00000241860.7</td>\n",
       "      <td>ENST00000662089.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>165889</td>\n",
       "      <td>168767</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000491962.1;Parent=ENSG00000241860.7;...</td>\n",
       "      <td>ENSG00000241860.7</td>\n",
       "      <td>ENST00000491962.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom  source     feature   start     end score strand frame  \\\n",
       "0  chr1  HAVANA  transcript  141474  149707     .      -     .   \n",
       "1  chr1  HAVANA  transcript  142808  146831     .      -     .   \n",
       "2  chr1  HAVANA  transcript  146386  173862     .      -     .   \n",
       "3  chr1  HAVANA  transcript  165491  169210     .      -     .   \n",
       "4  chr1  HAVANA  transcript  165889  168767     .      -     .   \n",
       "\n",
       "                                          attributes             Parent  \\\n",
       "0  ID=ENST00000484859.1;Parent=ENSG00000241860.7;...  ENSG00000241860.7   \n",
       "1  ID=ENST00000490997.5;Parent=ENSG00000241860.7;...  ENSG00000241860.7   \n",
       "2  ID=ENST00000466557.6;Parent=ENSG00000241860.7;...  ENSG00000241860.7   \n",
       "3  ID=ENST00000662089.1;Parent=ENSG00000241860.7;...  ENSG00000241860.7   \n",
       "4  ID=ENST00000491962.1;Parent=ENSG00000241860.7;...  ENSG00000241860.7   \n",
       "\n",
       "                 tID  \n",
       "0  ENST00000484859.1  \n",
       "1  ENST00000490997.5  \n",
       "2  ENST00000466557.6  \n",
       "3  ENST00000662089.1  \n",
       "4  ENST00000491962.1  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcripts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "4957a56c",
   "metadata": {},
   "outputs": [],
   "source": [
    "exons = processor.get_feature(genecode, 'exon')\n",
    "exons = pd.merge(exons, transcripts.tID, left_on='Parent', right_on='tID').drop('tID', axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5137f2ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_exons(gene, transcripts, exons):\n",
    "    gtrans= transcripts[transcripts.Parent == gene]\n",
    "    exons_list = [exons[exons.Parent==tr.tID].iloc[:, :5] for index, tr in gtrans.iterrows()]\n",
    "    return exons_list\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9db427a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_matrix(exons_list):\n",
    "    merged_df = bf.merge(pd.concat(exons_list), min_dist=0).drop('n_intervals', axis=1)\n",
    "    final_overlap = merged_df\n",
    "    for i, exon_df in enumerate(exons_list):\n",
    "        current_overlap = bf.overlap(final_overlap, exon_df)\n",
    "        current_overlap = current_overlap[~current_overlap.start.duplicated()]\n",
    "        matches = ~current_overlap['start_'].isna()\n",
    "        current_overlap = current_overlap.iloc[:, np.arange(0, 3+i)]\n",
    "        current_overlap['exon_'+str(i)] = 0\n",
    "        current_overlap.loc[matches, 'exon_'+str(i)] =1\n",
    "        final_overlap = current_overlap\n",
    "\n",
    "    final_overlap.iloc[:, 3:] = final_overlap.iloc[:, 3:].astype(int)\n",
    "    return final_overlap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "39380477",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "      <th>Description</th>\n",
       "      <th>Name</th>\n",
       "      <th>Parent</th>\n",
       "      <th>counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1109</th>\n",
       "      <td>chr2</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>217730</td>\n",
       "      <td>266398</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000035115.22;gene_id=ENSG00000035115....</td>\n",
       "      <td>SH3YL1</td>\n",
       "      <td>ENSG00000035115.22</td>\n",
       "      <td>ENSG00000035115.22</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1110</th>\n",
       "      <td>chr2</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>264140</td>\n",
       "      <td>278283</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000143727.16;gene_id=ENSG00000143727....</td>\n",
       "      <td>ACP1</td>\n",
       "      <td>ENSG00000143727.16</td>\n",
       "      <td>ENSG00000143727.16</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1111</th>\n",
       "      <td>chr2</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>279558</td>\n",
       "      <td>288851</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000189292.16;gene_id=ENSG00000189292....</td>\n",
       "      <td>ALKAL2</td>\n",
       "      <td>ENSG00000189292.16</td>\n",
       "      <td>ENSG00000189292.16</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1112</th>\n",
       "      <td>chr2</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>290941</td>\n",
       "      <td>314373</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000235779.8;gene_id=ENSG00000235779.8...</td>\n",
       "      <td>AC079779.5</td>\n",
       "      <td>ENSG00000235779.8</td>\n",
       "      <td>ENSG00000235779.8</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1113</th>\n",
       "      <td>chr2</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>558153</td>\n",
       "      <td>578145</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000233633.2;gene_id=ENSG00000233633.2...</td>\n",
       "      <td>AC093326.3</td>\n",
       "      <td>ENSG00000233633.2</td>\n",
       "      <td>ENSG00000233633.2</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1994</th>\n",
       "      <td>chr2</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>241675747</td>\n",
       "      <td>241686944</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000168393.13;gene_id=ENSG00000168393....</td>\n",
       "      <td>DTYMK</td>\n",
       "      <td>ENSG00000168393.13</td>\n",
       "      <td>ENSG00000168393.13</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1995</th>\n",
       "      <td>chr2</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>241687085</td>\n",
       "      <td>241729478</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000168395.16;gene_id=ENSG00000168395....</td>\n",
       "      <td>ING5</td>\n",
       "      <td>ENSG00000168395.16</td>\n",
       "      <td>ENSG00000168395.16</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1996</th>\n",
       "      <td>chr2</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>241734602</td>\n",
       "      <td>241768816</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000180902.18;gene_id=ENSG00000180902....</td>\n",
       "      <td>D2HGDH</td>\n",
       "      <td>ENSG00000180902.18</td>\n",
       "      <td>ENSG00000180902.18</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1997</th>\n",
       "      <td>chr2</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>241809065</td>\n",
       "      <td>241817413</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000204099.12;gene_id=ENSG00000204099....</td>\n",
       "      <td>NEU4</td>\n",
       "      <td>ENSG00000204099.12</td>\n",
       "      <td>ENSG00000204099.12</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1998</th>\n",
       "      <td>chr2</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>242088633</td>\n",
       "      <td>242175997</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000220804.9;gene_id=ENSG00000220804.9...</td>\n",
       "      <td>LINC01881</td>\n",
       "      <td>ENSG00000220804.9</td>\n",
       "      <td>ENSG00000220804.9</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>890 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom  source feature      start        end score strand frame  \\\n",
       "1109  chr2  HAVANA    gene     217730     266398     .      -     .   \n",
       "1110  chr2  HAVANA    gene     264140     278283     .      +     .   \n",
       "1111  chr2  HAVANA    gene     279558     288851     .      -     .   \n",
       "1112  chr2  HAVANA    gene     290941     314373     .      -     .   \n",
       "1113  chr2  HAVANA    gene     558153     578145     .      +     .   \n",
       "...    ...     ...     ...        ...        ...   ...    ...   ...   \n",
       "1994  chr2  HAVANA    gene  241675747  241686944     .      -     .   \n",
       "1995  chr2  HAVANA    gene  241687085  241729478     .      +     .   \n",
       "1996  chr2  HAVANA    gene  241734602  241768816     .      +     .   \n",
       "1997  chr2  HAVANA    gene  241809065  241817413     .      +     .   \n",
       "1998  chr2  HAVANA    gene  242088633  242175997     .      +     .   \n",
       "\n",
       "                                             attributes Description  \\\n",
       "1109  ID=ENSG00000035115.22;gene_id=ENSG00000035115....      SH3YL1   \n",
       "1110  ID=ENSG00000143727.16;gene_id=ENSG00000143727....        ACP1   \n",
       "1111  ID=ENSG00000189292.16;gene_id=ENSG00000189292....      ALKAL2   \n",
       "1112  ID=ENSG00000235779.8;gene_id=ENSG00000235779.8...  AC079779.5   \n",
       "1113  ID=ENSG00000233633.2;gene_id=ENSG00000233633.2...  AC093326.3   \n",
       "...                                                 ...         ...   \n",
       "1994  ID=ENSG00000168393.13;gene_id=ENSG00000168393....       DTYMK   \n",
       "1995  ID=ENSG00000168395.16;gene_id=ENSG00000168395....        ING5   \n",
       "1996  ID=ENSG00000180902.18;gene_id=ENSG00000180902....      D2HGDH   \n",
       "1997  ID=ENSG00000204099.12;gene_id=ENSG00000204099....        NEU4   \n",
       "1998  ID=ENSG00000220804.9;gene_id=ENSG00000220804.9...   LINC01881   \n",
       "\n",
       "                    Name              Parent  counts  \n",
       "1109  ENSG00000035115.22  ENSG00000035115.22      29  \n",
       "1110  ENSG00000143727.16  ENSG00000143727.16      13  \n",
       "1111  ENSG00000189292.16  ENSG00000189292.16       8  \n",
       "1112   ENSG00000235779.8   ENSG00000235779.8      39  \n",
       "1113   ENSG00000233633.2   ENSG00000233633.2       6  \n",
       "...                  ...                 ...     ...  \n",
       "1994  ENSG00000168393.13  ENSG00000168393.13       7  \n",
       "1995  ENSG00000168395.16  ENSG00000168395.16      13  \n",
       "1996  ENSG00000180902.18  ENSG00000180902.18      16  \n",
       "1997  ENSG00000204099.12  ENSG00000204099.12      17  \n",
       "1998   ENSG00000220804.9   ENSG00000220804.9      10  \n",
       "\n",
       "[890 rows x 13 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bf.select(genes, 'chr2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "33c1be9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "dummy_gene = 'ENSG00000180902.18'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "8148badc",
   "metadata": {},
   "outputs": [],
   "source": [
    "exons_list = get_exons(dummy_gene, transcripts, exons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "02ab4168",
   "metadata": {},
   "outputs": [],
   "source": [
    "exons_matrix = build_matrix(exons_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "bdb3c11f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>exon_0</th>\n",
       "      <th>exon_1</th>\n",
       "      <th>exon_2</th>\n",
       "      <th>exon_3</th>\n",
       "      <th>exon_4</th>\n",
       "      <th>exon_5</th>\n",
       "      <th>exon_6</th>\n",
       "      <th>exon_7</th>\n",
       "      <th>exon_8</th>\n",
       "      <th>exon_9</th>\n",
       "      <th>exon_10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr14</td>\n",
       "      <td>20343615</td>\n",
       "      <td>20343687</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr14</td>\n",
       "      <td>20344932</td>\n",
       "      <td>20345126</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr14</td>\n",
       "      <td>20345394</td>\n",
       "      <td>20345464</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr14</td>\n",
       "      <td>20346863</td>\n",
       "      <td>20346913</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr14</td>\n",
       "      <td>20350526</td>\n",
       "      <td>20350622</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr14</td>\n",
       "      <td>20350699</td>\n",
       "      <td>20351122</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr14</td>\n",
       "      <td>20352245</td>\n",
       "      <td>20352347</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr14</td>\n",
       "      <td>20352513</td>\n",
       "      <td>20353003</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr14</td>\n",
       "      <td>20354085</td>\n",
       "      <td>20354247</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr14</td>\n",
       "      <td>20354642</td>\n",
       "      <td>20354947</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr14</td>\n",
       "      <td>20355161</td>\n",
       "      <td>20355815</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr14</td>\n",
       "      <td>20355897</td>\n",
       "      <td>20356031</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr14</td>\n",
       "      <td>20356307</td>\n",
       "      <td>20356750</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr14</td>\n",
       "      <td>20357051</td>\n",
       "      <td>20357904</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom     start       end  exon_0  exon_1  exon_2  exon_3  exon_4  exon_5  \\\n",
       "0   chr14  20343615  20343687       1       1       1       1       0       0   \n",
       "1   chr14  20344932  20345126       1       1       1       1       0       0   \n",
       "2   chr14  20345394  20345464       1       1       1       1       0       0   \n",
       "3   chr14  20346863  20346913       1       1       1       1       0       0   \n",
       "4   chr14  20350526  20350622       1       0       1       1       0       0   \n",
       "5   chr14  20350699  20351122       1       0       1       1       1       0   \n",
       "6   chr14  20352245  20352347       1       1       1       1       1       0   \n",
       "7   chr14  20352513  20353003       0       0       0       0       0       1   \n",
       "8   chr14  20354085  20354247       1       1       1       1       1       1   \n",
       "9   chr14  20354642  20354947       1       1       1       1       0       0   \n",
       "10  chr14  20355161  20355815       1       0       1       1       0       0   \n",
       "11  chr14  20355897  20356031       1       0       1       1       0       0   \n",
       "12  chr14  20356307  20356750       1       0       1       1       0       0   \n",
       "13  chr14  20357051  20357904       1       0       1       1       0       0   \n",
       "\n",
       "    exon_6  exon_7  exon_8  exon_9  exon_10  \n",
       "0        0       0       0       0        0  \n",
       "1        0       0       0       0        0  \n",
       "2        0       0       0       0        0  \n",
       "3        0       0       0       0        0  \n",
       "4        0       0       0       0        0  \n",
       "5        0       0       0       0        0  \n",
       "6        0       0       0       0        0  \n",
       "7        0       0       0       0        0  \n",
       "8        0       0       0       0        0  \n",
       "9        1       0       0       0        0  \n",
       "10       1       1       1       1        0  \n",
       "11       1       1       1       1        1  \n",
       "12       0       1       1       1        1  \n",
       "13       0       0       1       1        0  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "810907f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "exons_matrix.to_csv(processed_path + dummy_gene+ \".tsv\", sep='\\t', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdb0e9a0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
