{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "51736a96",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from glob import glob\n",
    "import processor\n",
    "import bioframe as bf\n",
    "import numpy as np\n",
    "import bioframe.vis\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7b4738c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = '/scratch/groups/horence/angelika/EM_expression/data/'\n",
    "processed_path = './processed_files/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c73149c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/scratch/groups/horence/angelika/EM_expression/data/iso',\n",
       " '/scratch/groups/horence/angelika/EM_expression/data/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_exon_reads.parquet',\n",
       " '/scratch/groups/horence/angelika/EM_expression/data/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_reads.gct.gz',\n",
       " '/scratch/groups/horence/angelika/EM_expression/data/gencode.v38.annotation.gff3.gz',\n",
       " '/scratch/groups/horence/angelika/EM_expression/data/GTEx_Analysis_2017-06-05_v8_STARv2.5.3a_junctions.gct.gz']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glob(data_path+'*')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d7d41ae3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000223972.5;gene_id=ENSG00000223972.5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000456328.2;Parent=ENSG00000223972.5;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:1;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:2;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:3;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom  source     feature  start    end score strand frame  \\\n",
       "0  chr1  HAVANA        gene  11869  14409     .      +     .   \n",
       "1  chr1  HAVANA  transcript  11869  14409     .      +     .   \n",
       "2  chr1  HAVANA        exon  11869  12227     .      +     .   \n",
       "3  chr1  HAVANA        exon  12613  12721     .      +     .   \n",
       "4  chr1  HAVANA        exon  13221  14409     .      +     .   \n",
       "\n",
       "                                          attributes  \n",
       "0  ID=ENSG00000223972.5;gene_id=ENSG00000223972.5...  \n",
       "1  ID=ENST00000456328.2;Parent=ENSG00000223972.5;...  \n",
       "2  ID=exon:ENST00000456328.2:1;Parent=ENST0000045...  \n",
       "3  ID=exon:ENST00000456328.2:2;Parent=ENST0000045...  \n",
       "4  ID=exon:ENST00000456328.2:3;Parent=ENST0000045...  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genecode_path = data_path+'gencode.v38.annotation.gff3.gz'\n",
    "genecode = bf.read_table(genecode_path, schema='gff', skiprows=7)\n",
    "genecode = genecode[~genecode.start.isna()]\n",
    "genecode.start = genecode.start.astype(int)\n",
    "genecode.end = genecode.end.astype(int)\n",
    "\n",
    "genecode.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ce209704",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['./processed_files/genes_filtered.tsv']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glob(processed_path+'*')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "60a75918",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = pd.read_csv(processed_path+'genes_filtered.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "718dbab2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "      <th>Description</th>\n",
       "      <th>Name</th>\n",
       "      <th>Parent</th>\n",
       "      <th>counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>141474</td>\n",
       "      <td>173862</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000241860.7;gene_id=ENSG00000241860.7...</td>\n",
       "      <td>RP11-34P13.13</td>\n",
       "      <td>ENSG00000241860.7</td>\n",
       "      <td>ENSG00000241860.7</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>257864</td>\n",
       "      <td>359681</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000228463.10;gene_id=ENSG00000228463....</td>\n",
       "      <td>AP006222.1</td>\n",
       "      <td>ENSG00000228463.10</td>\n",
       "      <td>ENSG00000228463.10</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>365389</td>\n",
       "      <td>522928</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000237094.12;gene_id=ENSG00000237094....</td>\n",
       "      <td>RP4-669L17.4</td>\n",
       "      <td>ENSG00000237094.12</td>\n",
       "      <td>ENSG00000237094.12</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>586071</td>\n",
       "      <td>827796</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000230021.10;gene_id=ENSG00000230021....</td>\n",
       "      <td>RP11-206L10.17</td>\n",
       "      <td>ENSG00000230021.10</td>\n",
       "      <td>ENSG00000230021.10</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>778747</td>\n",
       "      <td>810065</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000237491.10;gene_id=ENSG00000237491....</td>\n",
       "      <td>LINC01409</td>\n",
       "      <td>ENSG00000237491.10</td>\n",
       "      <td>ENSG00000237491.10</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom  source feature   start     end score strand frame  \\\n",
       "0  chr1  HAVANA    gene  141474  173862     .      -     .   \n",
       "1  chr1  HAVANA    gene  257864  359681     .      -     .   \n",
       "2  chr1  HAVANA    gene  365389  522928     .      -     .   \n",
       "3  chr1  HAVANA    gene  586071  827796     .      -     .   \n",
       "4  chr1  HAVANA    gene  778747  810065     .      +     .   \n",
       "\n",
       "                                          attributes     Description  \\\n",
       "0  ID=ENSG00000241860.7;gene_id=ENSG00000241860.7...   RP11-34P13.13   \n",
       "1  ID=ENSG00000228463.10;gene_id=ENSG00000228463....      AP006222.1   \n",
       "2  ID=ENSG00000237094.12;gene_id=ENSG00000237094....    RP4-669L17.4   \n",
       "3  ID=ENSG00000230021.10;gene_id=ENSG00000230021....  RP11-206L10.17   \n",
       "4  ID=ENSG00000237491.10;gene_id=ENSG00000237491....       LINC01409   \n",
       "\n",
       "                 Name              Parent  counts  \n",
       "0   ENSG00000241860.7   ENSG00000241860.7       6  \n",
       "1  ENSG00000228463.10  ENSG00000228463.10       6  \n",
       "2  ENSG00000237094.12  ENSG00000237094.12      25  \n",
       "3  ENSG00000230021.10  ENSG00000230021.10      15  \n",
       "4  ENSG00000237491.10  ENSG00000237491.10      32  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "876d6a28",
   "metadata": {},
   "outputs": [],
   "source": [
    "transcripts = processor.get_feature(genecode, 'transcript')\n",
    "transcripts = pd.merge(transcripts, genes.Name, left_on='Parent', right_on='Name').drop('Name', axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "acad3167",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "      <th>Parent</th>\n",
       "      <th>tID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>141474</td>\n",
       "      <td>149707</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000484859.1;Parent=ENSG00000241860.7;...</td>\n",
       "      <td>ENSG00000241860.7</td>\n",
       "      <td>ENST00000484859.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>142808</td>\n",
       "      <td>146831</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000490997.5;Parent=ENSG00000241860.7;...</td>\n",
       "      <td>ENSG00000241860.7</td>\n",
       "      <td>ENST00000490997.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>146386</td>\n",
       "      <td>173862</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000466557.6;Parent=ENSG00000241860.7;...</td>\n",
       "      <td>ENSG00000241860.7</td>\n",
       "      <td>ENST00000466557.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>165491</td>\n",
       "      <td>169210</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000662089.1;Parent=ENSG00000241860.7;...</td>\n",
       "      <td>ENSG00000241860.7</td>\n",
       "      <td>ENST00000662089.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>165889</td>\n",
       "      <td>168767</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000491962.1;Parent=ENSG00000241860.7;...</td>\n",
       "      <td>ENSG00000241860.7</td>\n",
       "      <td>ENST00000491962.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom  source     feature   start     end score strand frame  \\\n",
       "0  chr1  HAVANA  transcript  141474  149707     .      -     .   \n",
       "1  chr1  HAVANA  transcript  142808  146831     .      -     .   \n",
       "2  chr1  HAVANA  transcript  146386  173862     .      -     .   \n",
       "3  chr1  HAVANA  transcript  165491  169210     .      -     .   \n",
       "4  chr1  HAVANA  transcript  165889  168767     .      -     .   \n",
       "\n",
       "                                          attributes             Parent  \\\n",
       "0  ID=ENST00000484859.1;Parent=ENSG00000241860.7;...  ENSG00000241860.7   \n",
       "1  ID=ENST00000490997.5;Parent=ENSG00000241860.7;...  ENSG00000241860.7   \n",
       "2  ID=ENST00000466557.6;Parent=ENSG00000241860.7;...  ENSG00000241860.7   \n",
       "3  ID=ENST00000662089.1;Parent=ENSG00000241860.7;...  ENSG00000241860.7   \n",
       "4  ID=ENST00000491962.1;Parent=ENSG00000241860.7;...  ENSG00000241860.7   \n",
       "\n",
       "                 tID  \n",
       "0  ENST00000484859.1  \n",
       "1  ENST00000490997.5  \n",
       "2  ENST00000466557.6  \n",
       "3  ENST00000662089.1  \n",
       "4  ENST00000491962.1  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcripts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "4957a56c",
   "metadata": {},
   "outputs": [],
   "source": [
    "exons = processor.get_feature(genecode, 'exon')\n",
    "exons = pd.merge(exons, transcripts.tID, left_on='Parent', right_on='tID').drop('tID', axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5137f2ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_exons(gene, transcripts, exons):\n",
    "    gtrans= transcripts[transcripts.Parent == gene]\n",
    "    exons_list = [exons[exons.Parent==tr.tID].iloc[:, :5] for index, tr in gtrans.iterrows()]\n",
    "    return exons_list\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9db427a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_matrix(exons_list):\n",
    "    merged_df = bf.merge(pd.concat(exons_list), min_dist=0).drop('n_intervals', axis=1)\n",
    "    final_overlap = merged_df\n",
    "    for i, exon_df in enumerate(exons_list):\n",
    "        current_overlap = bf.overlap(final_overlap, exon_df)\n",
    "        current_overlap = current_overlap[~current_overlap.start.duplicated()]\n",
    "        matches = ~current_overlap['start_'].isna()\n",
    "        current_overlap = current_overlap.iloc[:, np.arange(0, 3+i)]\n",
    "        current_overlap['exon_'+str(i)] = 0\n",
    "        current_overlap.loc[matches, 'exon_'+str(i)] =1\n",
    "        final_overlap = current_overlap\n",
    "\n",
    "    final_overlap.iloc[:, 3:] = final_overlap.iloc[:, 3:].astype(int)\n",
    "    return final_overlap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "33c1be9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "dummy_gene = 'ENSG00000228463.10'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "8148badc",
   "metadata": {},
   "outputs": [],
   "source": [
    "exons_list = get_exons(dummy_gene, transcripts, exons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "02ab4168",
   "metadata": {},
   "outputs": [],
   "source": [
    "exons_matrix = build_matrix(exons_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "bdb3c11f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>exon_0</th>\n",
       "      <th>exon_1</th>\n",
       "      <th>exon_2</th>\n",
       "      <th>exon_3</th>\n",
       "      <th>exon_4</th>\n",
       "      <th>exon_5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>257864</td>\n",
       "      <td>259025</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>261550</td>\n",
       "      <td>261634</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>263015</td>\n",
       "      <td>268204</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>268667</td>\n",
       "      <td>268816</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>289266</td>\n",
       "      <td>289370</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>297345</td>\n",
       "      <td>297502</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>356685</td>\n",
       "      <td>357586</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>358049</td>\n",
       "      <td>358183</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>359345</td>\n",
       "      <td>359681</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom   start     end  exon_0  exon_1  exon_2  exon_3  exon_4  exon_5\n",
       "0  chr1  257864  259025       1       1       1       1       1       0\n",
       "1  chr1  261550  261634       0       1       0       0       0       0\n",
       "2  chr1  263015  268204       1       1       0       1       0       1\n",
       "3  chr1  268667  268816       0       1       0       1       0       1\n",
       "4  chr1  289266  289370       0       0       0       0       0       1\n",
       "5  chr1  297345  297502       0       0       0       0       0       1\n",
       "6  chr1  356685  357586       0       0       1       0       0       0\n",
       "7  chr1  358049  358183       0       0       1       0       0       0\n",
       "8  chr1  359345  359681       0       0       1       0       0       0"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "810907f6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
