{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b535ec40",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import bioframe as bf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "dfec3509",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = '/scratch/groups/horence/angelika/EM_expression/data/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "26a5b7cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "genecode_path = data_path+'gencode.v38.annotation.gff3.gz'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7f2a975a",
   "metadata": {},
   "outputs": [],
   "source": [
    "gencode = bf.read_table(genecode_path, schema='gff', skiprows=7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "30973f26",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "      <th>Description</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000223972.5;gene_id=ENSG00000223972.5...</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000227232.5;gene_id=ENSG00000227232.5...</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000278267.1;gene_id=ENSG00000278267.1...</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000243485.5;gene_id=ENSG00000243485.5...</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000284332.1;gene_id=ENSG00000284332.1...</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom   source feature  start    end score strand frame  \\\n",
       "0   chr1   HAVANA    gene  11869  14409     .      +     .   \n",
       "12  chr1   HAVANA    gene  14404  29570     .      -     .   \n",
       "25  chr1  ENSEMBL    gene  17369  17436     .      -     .   \n",
       "28  chr1   HAVANA    gene  29554  31109     .      +     .   \n",
       "36  chr1  ENSEMBL    gene  30366  30503     .      +     .   \n",
       "\n",
       "                                           attributes  Description  \\\n",
       "0   ID=ENSG00000223972.5;gene_id=ENSG00000223972.5...      DDX11L1   \n",
       "12  ID=ENSG00000227232.5;gene_id=ENSG00000227232.5...       WASH7P   \n",
       "25  ID=ENSG00000278267.1;gene_id=ENSG00000278267.1...    MIR6859-1   \n",
       "28  ID=ENSG00000243485.5;gene_id=ENSG00000243485.5...  MIR1302-2HG   \n",
       "36  ID=ENSG00000284332.1;gene_id=ENSG00000284332.1...    MIR1302-2   \n",
       "\n",
       "                 Name  \n",
       "0   ENSG00000223972.5  \n",
       "12  ENSG00000227232.5  \n",
       "25  ENSG00000278267.1  \n",
       "28  ENSG00000243485.5  \n",
       "36  ENSG00000284332.1  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = gencode.loc[gencode['feature'].values=='gene']\n",
    "genes = genes.astype({'start':int,'end':int})\n",
    "### need to join gene table & expression table\n",
    "geneNames =  genes['attributes'].str.split(';')\n",
    "genes['Description'] = geneNames.str[3].str.split('=').str[1]\n",
    "genes['Name'] = geneNames.str[0].str.split('=').str[1]\n",
    "genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3cb5e06a",
   "metadata": {},
   "outputs": [],
   "source": [
    "trans_df = gencode[(gencode.feature == 'transcript')].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "54cc994c",
   "metadata": {},
   "outputs": [],
   "source": [
    "trans_df['Parent'] = trans_df.attributes.str.split(';').str[1].str.split('=').str[1]\n",
    "trans_df['tID'] = trans_df.attributes.str.split(';').str[0].str.split('=').str[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "34cfd09a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "      <th>Parent</th>\n",
       "      <th>tID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869.0</td>\n",
       "      <td>14409.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000456328.2;Parent=ENSG00000223972.5;...</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>12010.0</td>\n",
       "      <td>13670.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000450305.2;Parent=ENSG00000223972.5;...</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14404.0</td>\n",
       "      <td>29570.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000488147.1;Parent=ENSG00000227232.5;...</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>17369.0</td>\n",
       "      <td>17436.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000619216.1;Parent=ENSG00000278267.1;...</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>ENST00000619216.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>29554.0</td>\n",
       "      <td>31097.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000473358.1;Parent=ENSG00000243485.5;...</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>ENST00000473358.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3148142</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14149.0</td>\n",
       "      <td>14673.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000361681.2;Parent=ENSG00000198695.2;...</td>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>ENST00000361681.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3148147</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14674.0</td>\n",
       "      <td>14742.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000387459.1;Parent=ENSG00000210194.1;...</td>\n",
       "      <td>ENSG00000210194.1</td>\n",
       "      <td>ENST00000387459.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3148150</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14747.0</td>\n",
       "      <td>15887.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000361789.2;Parent=ENSG00000198727.2;...</td>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>ENST00000361789.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3148155</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15888.0</td>\n",
       "      <td>15953.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000387460.2;Parent=ENSG00000210195.2;...</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>ENST00000387460.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3148158</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15956.0</td>\n",
       "      <td>16023.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000387461.2;Parent=ENSG00000210196.2;...</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>ENST00000387461.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>237012 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom   source     feature    start      end score strand frame  \\\n",
       "1        chr1   HAVANA  transcript  11869.0  14409.0     .      +     .   \n",
       "5        chr1   HAVANA  transcript  12010.0  13670.0     .      +     .   \n",
       "13       chr1   HAVANA  transcript  14404.0  29570.0     .      -     .   \n",
       "26       chr1  ENSEMBL  transcript  17369.0  17436.0     .      -     .   \n",
       "29       chr1   HAVANA  transcript  29554.0  31097.0     .      +     .   \n",
       "...       ...      ...         ...      ...      ...   ...    ...   ...   \n",
       "3148142  chrM  ENSEMBL  transcript  14149.0  14673.0     .      -     .   \n",
       "3148147  chrM  ENSEMBL  transcript  14674.0  14742.0     .      -     .   \n",
       "3148150  chrM  ENSEMBL  transcript  14747.0  15887.0     .      +     .   \n",
       "3148155  chrM  ENSEMBL  transcript  15888.0  15953.0     .      +     .   \n",
       "3148158  chrM  ENSEMBL  transcript  15956.0  16023.0     .      -     .   \n",
       "\n",
       "                                                attributes             Parent  \\\n",
       "1        ID=ENST00000456328.2;Parent=ENSG00000223972.5;...  ENSG00000223972.5   \n",
       "5        ID=ENST00000450305.2;Parent=ENSG00000223972.5;...  ENSG00000223972.5   \n",
       "13       ID=ENST00000488147.1;Parent=ENSG00000227232.5;...  ENSG00000227232.5   \n",
       "26       ID=ENST00000619216.1;Parent=ENSG00000278267.1;...  ENSG00000278267.1   \n",
       "29       ID=ENST00000473358.1;Parent=ENSG00000243485.5;...  ENSG00000243485.5   \n",
       "...                                                    ...                ...   \n",
       "3148142  ID=ENST00000361681.2;Parent=ENSG00000198695.2;...  ENSG00000198695.2   \n",
       "3148147  ID=ENST00000387459.1;Parent=ENSG00000210194.1;...  ENSG00000210194.1   \n",
       "3148150  ID=ENST00000361789.2;Parent=ENSG00000198727.2;...  ENSG00000198727.2   \n",
       "3148155  ID=ENST00000387460.2;Parent=ENSG00000210195.2;...  ENSG00000210195.2   \n",
       "3148158  ID=ENST00000387461.2;Parent=ENSG00000210196.2;...  ENSG00000210196.2   \n",
       "\n",
       "                       tID  \n",
       "1        ENST00000456328.2  \n",
       "5        ENST00000450305.2  \n",
       "13       ENST00000488147.1  \n",
       "26       ENST00000619216.1  \n",
       "29       ENST00000473358.1  \n",
       "...                    ...  \n",
       "3148142  ENST00000361681.2  \n",
       "3148147  ENST00000387459.1  \n",
       "3148150  ENST00000361789.2  \n",
       "3148155  ENST00000387460.2  \n",
       "3148158  ENST00000387461.2  \n",
       "\n",
       "[237012 rows x 11 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trans_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "afee5fd7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Parent</th>\n",
       "      <th>counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003.15</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005.6</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457.14</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Parent  counts\n",
       "0  ENSG00000000003.15       5\n",
       "1   ENSG00000000005.6       2\n",
       "2  ENSG00000000419.14      16\n",
       "3  ENSG00000000457.14       5\n",
       "4  ENSG00000000460.17       9"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trans_counts = trans_df.groupby('Parent').size().reset_index(name='counts')\n",
    "trans_counts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f4813202",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes =  pd.merge(genes, trans_counts, left_on='Name', right_on='Parent')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "ff8b5ee8",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_filtered = genes[genes.counts > 5].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "d643a6da",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_filtered.to_csv('./processed_files/genes_filtered.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "7c8e9c27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "      <th>Description</th>\n",
       "      <th>Name</th>\n",
       "      <th>Parent</th>\n",
       "      <th>counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>141474</td>\n",
       "      <td>173862</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000241860.7;gene_id=ENSG00000241860.7...</td>\n",
       "      <td>RP11-34P13.13</td>\n",
       "      <td>ENSG00000241860.7</td>\n",
       "      <td>ENSG00000241860.7</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>257864</td>\n",
       "      <td>359681</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000228463.10;gene_id=ENSG00000228463....</td>\n",
       "      <td>AP006222.1</td>\n",
       "      <td>ENSG00000228463.10</td>\n",
       "      <td>ENSG00000228463.10</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>365389</td>\n",
       "      <td>522928</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000237094.12;gene_id=ENSG00000237094....</td>\n",
       "      <td>RP4-669L17.4</td>\n",
       "      <td>ENSG00000237094.12</td>\n",
       "      <td>ENSG00000237094.12</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>586071</td>\n",
       "      <td>827796</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000230021.10;gene_id=ENSG00000230021....</td>\n",
       "      <td>RP11-206L10.17</td>\n",
       "      <td>ENSG00000230021.10</td>\n",
       "      <td>ENSG00000230021.10</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>778747</td>\n",
       "      <td>810065</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000237491.10;gene_id=ENSG00000237491....</td>\n",
       "      <td>LINC01409</td>\n",
       "      <td>ENSG00000237491.10</td>\n",
       "      <td>ENSG00000237491.10</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12201</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>20464916</td>\n",
       "      <td>20575519</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000229236.3;gene_id=ENSG00000229236.3...</td>\n",
       "      <td>TTTY10</td>\n",
       "      <td>ENSG00000229236.3</td>\n",
       "      <td>ENSG00000229236.3</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12202</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>23219434</td>\n",
       "      <td>23291356</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000205944.11;gene_id=ENSG00000205944....</td>\n",
       "      <td>DAZ2</td>\n",
       "      <td>ENSG00000205944.11</td>\n",
       "      <td>ENSG00000205944.11</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12203</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>24833843</td>\n",
       "      <td>24907040</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000205916.12;gene_id=ENSG00000205916....</td>\n",
       "      <td>DAZ4</td>\n",
       "      <td>ENSG00000205916.12</td>\n",
       "      <td>ENSG00000205916.12</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12204</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>57067865</td>\n",
       "      <td>57130289</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000124333.16_PAR_Y;gene_id=ENSG000001...</td>\n",
       "      <td>VAMP7</td>\n",
       "      <td>ENSG00000124333.16_PAR_Y</td>\n",
       "      <td>ENSG00000124333.16_PAR_Y</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12205</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>57207346</td>\n",
       "      <td>57212230</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000182484.15_PAR_Y;gene_id=ENSG000001...</td>\n",
       "      <td>WASH6P</td>\n",
       "      <td>ENSG00000182484.15_PAR_Y</td>\n",
       "      <td>ENSG00000182484.15_PAR_Y</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12206 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom  source feature     start       end score strand frame  \\\n",
       "0      chr1  HAVANA    gene    141474    173862     .      -     .   \n",
       "1      chr1  HAVANA    gene    257864    359681     .      -     .   \n",
       "2      chr1  HAVANA    gene    365389    522928     .      -     .   \n",
       "3      chr1  HAVANA    gene    586071    827796     .      -     .   \n",
       "4      chr1  HAVANA    gene    778747    810065     .      +     .   \n",
       "...     ...     ...     ...       ...       ...   ...    ...   ...   \n",
       "12201  chrY  HAVANA    gene  20464916  20575519     .      -     .   \n",
       "12202  chrY  HAVANA    gene  23219434  23291356     .      +     .   \n",
       "12203  chrY  HAVANA    gene  24833843  24907040     .      +     .   \n",
       "12204  chrY  HAVANA    gene  57067865  57130289     .      +     .   \n",
       "12205  chrY  HAVANA    gene  57207346  57212230     .      +     .   \n",
       "\n",
       "                                              attributes     Description  \\\n",
       "0      ID=ENSG00000241860.7;gene_id=ENSG00000241860.7...   RP11-34P13.13   \n",
       "1      ID=ENSG00000228463.10;gene_id=ENSG00000228463....      AP006222.1   \n",
       "2      ID=ENSG00000237094.12;gene_id=ENSG00000237094....    RP4-669L17.4   \n",
       "3      ID=ENSG00000230021.10;gene_id=ENSG00000230021....  RP11-206L10.17   \n",
       "4      ID=ENSG00000237491.10;gene_id=ENSG00000237491....       LINC01409   \n",
       "...                                                  ...             ...   \n",
       "12201  ID=ENSG00000229236.3;gene_id=ENSG00000229236.3...          TTTY10   \n",
       "12202  ID=ENSG00000205944.11;gene_id=ENSG00000205944....            DAZ2   \n",
       "12203  ID=ENSG00000205916.12;gene_id=ENSG00000205916....            DAZ4   \n",
       "12204  ID=ENSG00000124333.16_PAR_Y;gene_id=ENSG000001...           VAMP7   \n",
       "12205  ID=ENSG00000182484.15_PAR_Y;gene_id=ENSG000001...          WASH6P   \n",
       "\n",
       "                           Name                    Parent  counts  \n",
       "0             ENSG00000241860.7         ENSG00000241860.7       6  \n",
       "1            ENSG00000228463.10        ENSG00000228463.10       6  \n",
       "2            ENSG00000237094.12        ENSG00000237094.12      25  \n",
       "3            ENSG00000230021.10        ENSG00000230021.10      15  \n",
       "4            ENSG00000237491.10        ENSG00000237491.10      32  \n",
       "...                         ...                       ...     ...  \n",
       "12201         ENSG00000229236.3         ENSG00000229236.3      13  \n",
       "12202        ENSG00000205944.11        ENSG00000205944.11       9  \n",
       "12203        ENSG00000205916.12        ENSG00000205916.12       8  \n",
       "12204  ENSG00000124333.16_PAR_Y  ENSG00000124333.16_PAR_Y       6  \n",
       "12205  ENSG00000182484.15_PAR_Y  ENSG00000182484.15_PAR_Y      15  \n",
       "\n",
       "[12206 rows x 13 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "00428b48",
   "metadata": {},
   "outputs": [],
   "source": [
    "# exons = gencode[gencode.feature=='exon'].copy()\n",
    "# exons['Parent'] = exons.attributes.str.split(';').str[1].str.split('=').str[1]\n",
    "# exons['eID'] = exons.attributes.str.split(';').str[0].str.split('=').str[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "60bd78b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "ef71f63b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gtex_genes_path = data_path+'GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_reads.gct.gz'\n",
    "# gtex_junctions_path = data_path+'GTEx_Analysis_2017-06-05_v8_STARv2.5.3a_junctions.gct.gz'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "3536e35f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gtex_junctions = pd.read_csv(gtex_junctions_path, delimiter='\\t', skiprows = 2, nrows=1000, )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "63f4cdcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gtex_junctions[gtex_junctions.Description=='ENSG00000227232.5']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e572c82",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
